rustPlatform.fetchCargoVendor: init

TomaSajt opened the pull request:
Related: https://github.com/NixOS/nixpkgs/issues/327063

---

This PR adds an alternative method to vendor the dependencies of a cargo project.
It's currently implemented in python, but could be changed later, if need be.

I converted some packages to use this fetcher as an example.

If you can come up with a good name for this fetcher, please share it below.

If we use this fetcher we won't have to worry about cargo changing its vendoring implementation (since this is completely custom), thus we can replace all `importCargoLock` usages where it was only required because of the presence of git dependencies.

The main idea is that we try to create a minimal staging FOD.
We fetch the tarballs into `$out/tarballs/${name}-${version}.tar.gz` and we fetch the git repos to `$out/git/${git_sha_rev}`.

This is the most important part to get right, since we will only be able to minimally change this in the future.
We would still be able to add some extra behaviours in cases that currently just throw an error (e.g. fetching other registries).

---

The staging FOD will then be used to create the actual vendor dir.

Since this no longer requires network access, we can freely change the implementation of this part without worrying about that hash.

The cargo workspace inheritance logic was already reimplemented by a script written for `importCargoLock`, so I just reused that.

---

Since this is a two-step process and only the first step is a FOD, we won't have `outputHash` and friends on the result. If we want to override the `outputHash`, we would have to do so through the `vendorStaging` attribute.

TODO: it might be useful to add an easier way to override the `outputHash` of `vendorStaging`.

---

I added a commit that makes `buildCargoPackage` swap over to this new fetcher implementation. This is currently done through a boolean flag.
Should it maybe be an enum instead? e.g. `cargoFetcher = "custom"` vs `cargoFetcher = "default"`;

---

Sidenote:
The current vendoring script only requires the `Cargo.lock` file and nothing else. This would mean that `fetchCargoVendor` may easily be changed to not take in `src` and friends, but only a path to the cargo lock, like `${src}/Cargo.lock`. This is not IFD, since we never import the path from nix, only from a script. This would be similar to how it's done with `fetchYarnDeps`.

Having it as a normal mkDerivation, like it is now, we can also apply patches and use hooks, which might be preferred.

---

It would be possible to skip the staging phase and make the vendor dir the FOD, however, this would lock our output format to never be able to change again.

---

The current implementation uses `nix-prefetch-git` to fetch a reproducible git tree, as I wanted to make sure I didn't make any implementation mistakes. I wonder if there's a better method than this...

---

Nothing is stopping us from also compressing the final vendor directory into a tarball, just like with `fetchCargoTarball`.
Tarballs - obviously - take up less space, so ideally we would do this. 
Though a better fetcher name might be needed in that case.
I'm keeping the output as a directory for now, until the testing is done.

---

I also worry about the fact that non-FOD derivations will rebuild whenever the inputs change. This would mean that whenever `python` needs to be rebuilt, every vendor directory created by this would also need to be rebuilt. The rebuilds of the vendor directories would not take much time, but the used up space would be a concern.
It would be great if we could tell hydra not to cache certain derivations. 

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

TomaSajt committed:
rustPlatform.fetchCargoVendor: init

Sigmanificient commented on pkgs/build-support/rust/mk-cargo-vendor-deps-script.py:
```diff
@@ -0,0 +1,230 @@
+import glob
```
Should it has the nix shebang?
```suggestion
#! /usr/bin/env nix-shell
#! nix-shell -i python3 -p python3
```

Sigmanificient commented on pkgs/build-support/rust/mk-cargo-vendor-deps-script.py:
```diff
@@ -0,0 +1,223 @@
+            shutil.copytree(crate_tree, crate_out_dir)
+            os.chmod(crate_out_dir, 0o755)
+
+            with open(os.path.join(crate_out_dir, ".cargo-checksum.json"), "w") as f:
+                f.write('{"files":{},"package":null}')
```
This could be replaced by a call to `write_checksum_json`

TomaSajt replied:
This function was accidentally left behind

Sigmanificient commented on pkgs/build-support/rust/mk-cargo-vendor-deps-script.py:
```diff
@@ -0,0 +1,223 @@
+import glob
```
```suggestion
#! /usr/bin/env nix-shell
#! nix-shell -i python3 -p python3
import glob
```

TomaSajt committed:
hieroglypic: migrate to new fetcher

TomaSajt commented on pkgs/build-support/rust/mk-cargo-vendor-deps-script.py:
```diff
@@ -0,0 +1,223 @@
+
+# TODO: support other repositories
+#       or maybe allow getting downloadUrl through an API request
+def download_tarball(args: tuple[dict, str]):
```
I think this needed to be done in order to work with multi-threading.
I'll look into it if this is required.

TomaSajt replied:
Found out that `pool.starmap` does the unpacking automatically.

TomaSajt committed:
popsicle: migrate to new fetcher

TomaSajt committed:
python312Packages.json-stream-rs-tokenizer: migrate to new fetcher

TomaSajt commented:
@ofborg build granian popsicle

TomaSajt commented:
I wonder why ofborg is failing...

TomaSajt commented:
@ofborg build granian popsicle

TomaSajt commented on pkgs/build-support/rust/mk-cargo-vendor-deps-script.py:
```diff
@@ -0,0 +1,230 @@
+def download_tarball(args: tuple[dict, str]):
+    pkg, out_dir = args
+    downloadUrl = "https://crates.io/api/v1/crates"
+    url = f"{downloadUrl}/{pkg["name"]}/{pkg["version"]}/download"
```
Can be an improvement later

TomaSajt commented on pkgs/build-support/rust/mk-cargo-vendor-deps-script.py:
```diff
@@ -0,0 +1,223 @@
+    if matches is None:
+        raise Exception(f"Unable to process git source: {source}")
+    url, _, type_, value, git_sha_rev = matches.groups()
+    return {"url": url, "type": type_, "value": value, "git_sha_rev": git_sha_rev}
```
This is great

Sigmanificient commented:
What about using `black --max-length 79` to ensure pep8 compliance & flake8 E501 rule?

TomaSajt commented:
> What about using `black --max-length 79` to ensure pep8 compliance & flake8 E501 rule?
> It would also make the diff easier to read in split mode (as it is ~90 chars)

Maybe later.
Before that, I would like to see if there's even a demand for this new fetcher.

Sigmanificient commented:
> Maybe later. Before that, I would like to see if there's even a demand for this new fetcher.

No problem, I saw python and couldn't resist looking at it :eyes:
hope for the best!

Mic92 commented on pkgs/build-support/rust/fetch-cargo-vendor-util.py:
```diff
@@ -0,0 +1,247 @@
+        raise Exception(f"Hash mismatch! File fetched from {url} had checksum {calculated_checksum}, expected {expected_checksum}.")
+
+
+def get_download_url_for_tarball(pkg: dict):
```
```suggestion
def get_download_url_for_tarball(pkg: dict[str, Any]) -> str:
```

Sigmanificient replied:
I proposed to use `typeDict` above, which would provide way better typing

Mic92 commented on pkgs/build-support/rust/fetch-cargo-vendor-util.py:
```diff
@@ -0,0 +1,247 @@
+    return f"https://crates.io/api/v1/crates/{pkg["name"]}/{pkg["version"]}/download"
+
+
+def download_tarball(pkg: dict, out_dir: str):
```
```suggestion
def download_tarball(pkg: dict[str, Any], out_dir: str) -> None:
```

Mic92 commented on pkgs/build-support/rust/fetch-cargo-vendor-util.py:
```diff
@@ -0,0 +1,247 @@
+    download_file_with_checksum(url, tarball_out_dir, expected_checksum)
+
+
+def download_git_tree(url: str, git_sha_rev: str, out_dir: str):
```
```suggestion
def download_git_tree(url: str, git_sha_rev: str, out_dir: str) -> None:
```

Mic92 commented on pkgs/build-support/rust/fetch-cargo-vendor-util.py:
```diff
@@ -0,0 +1,247 @@
+    tree_out_dir = os.path.join(out_dir, "git", git_sha_rev)
+    eprint(f"Fetching {url}#{git_sha_rev} -> git/{git_sha_rev}")
+
+    subprocess.check_output(["nix-prefetch-git", "--builder", "--quiet", "--url", url, "--rev", git_sha_rev, "--out", tree_out_dir])
```
No need to capture output here:

```suggestion
    subprocess.run(["nix-prefetch-git", "--builder", "--quiet", "--url", url, "--rev", git_sha_rev, "--out", tree_out_dir], check=True)
```

Mic92 commented on pkgs/build-support/rust/fetch-cargo-vendor-util.py:
```diff
@@ -0,0 +1,247 @@
+GIT_SOURCE_REGEX = re.compile("git\\+(?P<url>[^?]+)(\\?(?P<type>rev|tag|branch)=(?P<value>.*))?#(?P<git_sha_rev>.*)")
+
+
+def parse_git_source(source: str):
```
Not sure about the return type here?

```suggestion
def parse_git_source(source: str) -> dict[str,str]
```

Mic92 commented on pkgs/build-support/rust/fetch-cargo-vendor-util.py:
```diff
@@ -0,0 +1,247 @@
+    return matches.groupdict()
+
+
+def create_vendor_staging(lockfile_path: str, out_dir: str):
```
```suggestion
def create_vendor_staging(lockfile_path: str, out_dir: str) -> None:
```

Mic92 commented on pkgs/build-support/rust/fetch-cargo-vendor-util.py:
```diff
@@ -0,0 +1,247 @@
+    os.makedirs(out_dir, exist_ok=True)
+    shutil.copy(lockfile_path, os.path.join(out_dir, "Cargo.lock"))
+
+    pool = mp.Pool(min(10, mp.cpu_count()))
+
+    if len(git_packages) != 0:
+        os.mkdir(os.path.join(out_dir, "git"))
+        # run download jobs in parallel
+        pool.starmap(download_git_tree, ((url, git_sha_rev, out_dir) for git_sha_rev, url in git_sha_rev_to_url.items()))
+
+    if len(registry_packages) != 0:
+        os.mkdir(os.path.join(out_dir, "tarballs"))
+        # run download jobs in parallel
+        pool.starmap(download_tarball, ((pkg, out_dir) for pkg in registry_packages))
+
+    pool.close()
```
```suggestion
    with mp.Pool(min(10, mp.cpu_count())) as pool:

      if len(git_packages) != 0:
          os.mkdir(os.path.join(out_dir, "git"))
          # run download jobs in parallel
          pool.starmap(download_git_tree, ((url, git_sha_rev, out_dir) for git_sha_rev, url in git_sha_rev_to_url.items()))

      if len(registry_packages) != 0:
          os.mkdir(os.path.join(out_dir, "tarballs"))
          # run download jobs in parallel
          pool.starmap(download_tarball, ((pkg, out_dir) for pkg in registry_packages))
```

Sigmanificient commented:
@Mic92  For typing dictionnary, I proposed to use `typedDict` above, which would provide way better typing

Mic92 commented on pkgs/build-support/rust/fetch-cargo-vendor-util.py:
```diff
@@ -0,0 +1,247 @@
+    pool.close()
+
+
+def get_manifest_metadata(manifest_path: str):
```
```suggestion
def get_manifest_metadata(manifest_path: str) -> dict[str, Any]
```

Mic92 commented on pkgs/build-support/rust/fetch-cargo-vendor-util.py:
```diff
@@ -0,0 +1,247 @@
+    return crate_manifest_path
+
+
+def find_crate_manifest_in_tree(tree: str, crate_name: str):
```
```suggestion
def find_crate_manifest_in_tree(tree: str, crate_name: str) -> str | None
```

Mic92 commented on pkgs/build-support/rust/fetch-cargo-vendor-util.py:
```diff
@@ -0,0 +1,247 @@
+
+def find_crate_manifest_in_tree(tree: str, crate_name: str):
+    # in some cases Cargo.toml is not located at the top level, so we also look at subdirectories
+    manifest_paths = (os.path.join(tree, path) for path in glob.iglob("**/Cargo.toml", root_dir=tree, recursive=True))
```
Consider using `pathlib` here.

Mic92 replied:
It adds some type safety and makes joining path a bit nicer and more portable.

TomaSajt replied:
Refactored the entire thing to use pathlib.

Mic92 commented on pkgs/build-support/rust/fetch-cargo-vendor-util.py:
```diff
@@ -0,0 +1,247 @@
+    raise Exception(f"Couldn't find manifest for crate {crate_name} inside {tree}.")
+
+
+def create_vendor(vendor_staging_dir: str, out_dir: str):
```
```suggestion
def create_vendor(vendor_staging_dir: str, out_dir: str) -> None:
```

Mic92 commented on pkgs/build-support/rust/fetch-cargo-vendor-util.py:
```diff
@@ -0,0 +1,247 @@
+    elif method == "create-vendor":
+        create_vendor(vendor_staging_dir=sys.argv[2], out_dir=sys.argv[3])
+    else:
+        raise Exception(f"Unknown method: {method}")
```
I would make create your own `Exception` at the start of the script and than catch this exception in the main method:

```
try:
    ...
exception Error as e:
   eprint(e)
```

This won't show than error stacktraces for our own errors, which is a bit more user friendly (because users often don't find the relevant error from a stack trace).

Mic92 replied:
```suggestion
        raise Exception(f"Unknown subcommand: {method} must be one of: create-vendor-staging, create-vendor")
```

Mic92 commented on pkgs/build-support/rust/fetch-cargo-vendor.nix:
```diff
@@ -0,0 +1,93 @@
+      fetchCargoVendorUtil
+      cargo
+      replaceWorkspaceValues
+    ];
```
Would be nice to also type check this:

```suggestion
    ];
    passthru.tests = {
      mypy = pkgs.runCommand "mypy-check" ''
      '';
    };
```

JohnRTitor replied:
@TomaSajt

Mic92 commented on pkgs/build-support/rust/fetch-cargo-vendor.nix:
```diff
@@ -0,0 +1,93 @@
+  replaceWorkspaceValues = writers.writePython3Bin "replace-workspace-values" {
+    libraries = with python3Packages; [
+      tomli
+      tomli-w
```
python also can parse toml nativly now. It's only needed for writing, which doesn't seem to be used here?

TomaSajt replied:
`tomli-w` is only used by `replace-workspace-values.py`, which I didn't write.
I migrated over to `tomllib` for the vendor script.

Mic92 commented:
> @Mic92 For typing dictionnaries, I proposed to use `typedDict` above, which would provide way better typing than having an `Any` as value type

ok.

Mic92 commented:
If this can replace our cargo.lock files, this is a great addition. I hope we can consolidate this and replace our current vendoring.

Mic92 commented:
> What about using `black --max-length 79` to ensure pep8 compliance & flake8 E501 rule? It would also make the diff easier to read in split mode (as it is ~90 chars)

I think we can just use ruff format and ruff check. Pretty much the same thing but also more potential lints and also faster.

Sigmanificient commented:
> > @Mic92 For typing dictionnaries, I proposed to use `typedDict` above, which would provide way better typing than having an `Any` as value type
> 
> ok with me. Does typedDict has to include all fields or only the ones we care about?

you can set `total=False` on the class to hint that the typing only has the relevant key

Sigmanificient commented:
> > What about using `black --max-length 79` to ensure pep8 compliance & flake8 E501 rule? It would also make the diff easier to read in split mode (as it is ~90 chars)
> 
> I think we can just use ruff format and ruff check. Pretty much the same thing but also more potential lints and also faster.

Yeah, i suggested black because i've been using it for years ahah, has long has it complies with pep8 i believe it's fine

Mic92 commented:
> > > What about using `black --max-length 79` to ensure pep8 compliance & flake8 E501 rule? It would also make the diff easier to read in split mode (as it is ~90 chars)
> > 
> > 
> > I think we can just use ruff format and ruff check. Pretty much the same thing but also more potential lints and also faster.
> 
> Yeah, i suggested black because i've been using it for years ahah, has long has it complies with pep8 i believe it's fine

ruff format should be almost 100% compatible to black.

SuperSandro2000 commented on pkgs/build-support/rust/fetch-cargo-vendor.nix:
```diff
@@ -0,0 +1,92 @@
+      dontFixup = true;
+
+      outputHash = hash;
+      outputHashAlgo = "sha256";
```
With hardcoding sha256 here, can we still use ``hash = sha512-....``.

TomaSajt commented on pkgs/build-support/rust/fetch-cargo-vendor-util.py:
```diff
@@ -0,0 +1,247 @@
+
+        source: str = pkg["source"]
+
+        if source.startswith("git+"):
```
I factored out a part of the logic,

TomaSajt commented:
I thought of an interesting alternative to what's being done currently in this PR.

Instead of having two derivations, the FOD and the one constructed from the FOD, we would only have the FOD.
The construction of the actual vendor dir would then be moved to be handled by `cargoSetupHook` (it would just call the util script with the `create-vendor` subcommand)

This might be a better solution, though I haven't decided on it yet.

Sigmanificient commented on pkgs/build-support/rust/fetch-cargo-vendor-util.py:
```diff
@@ -0,0 +1,278 @@
+    match = GIT_SOURCE_REGEX.match(source)
+    if match is None:
+        raise Exception(f"Unable to process git source: {source}.")
+    # hack: make the type system accept that the function outputs GitSourceInfo
+    source_info: Any = match.groupdict(default=None)
+    return source_info
```
`typing.cast` allow you to explicitely perform the same hack, in a cleaner way
```suggestion
    return cast(GitSourceInfo, match.groupdict(default=None))
```

lucastso10 commented:
This changes will not be relevant for the granian package when #348813 merges, since they dropped their git dependencies

TomaSajt commented:
> This changes will not be relevant for the granian package when #348813 merges, since they dropped their git dependencies

I see, thanks for the heads up. Just selected a few packages at random.

TomaSajt committed:
buildRustPackage: add useFetchCargoVendor flag

TomaSajt committed:
cargo-shuttle: migrate to new fetcher

Mic92 commented on pkgs/build-support/rust/fetch-cargo-vendor-util.py:
```diff
@@ -0,0 +1,284 @@
+    shutil.copy(lockfile_path, out_dir / "Cargo.lock")
+
+    # create a pool with at most 10 concurrent jobs
+    with mp.Pool(min(10, mp.cpu_count())) as pool:
```
Since this is IO based, I don't think we need to bind this to the number of cores. 
10 is a bit more than what browsers do (firefox does 10). Otherwise when I run this on an EPYC, I get 96 connections, which github might not like.

```suggestion
    with mp.Pool(10) as pool:
```

TomaSajt replied:
I used min(10,x) to clamp x to not be greater than 10.
I could change it to 7. Maybe I should look at cargo's implementation.

If we just use 10 or 7 without considering the core count will it know automatically not to try more things it can handle? I'll probably have to read the docs of multiprocessing to know for sure.

emilazy replied:
I think `min` was misread as `max`.

TomaSajt replied:
Looked up [the docs](https://docs.python.org/3/library/multiprocessing.html#module-multiprocessing.pool) and the default value for the number of processes seems to be [os.cpu_count()](https://docs.python.org/3/library/os.html#os.cpu_count) for `python<=3.12` and [os.process_cpu_count()](https://docs.python.org/3/library/os.html#os.process_cpu_count) for `python>=3.13`.

Based on this info, I would say that we should use the current clamping method.

(I'm assuming `multiprocessing.cpu_count()` is the same as `os.cpu_count()`)

Mic92 replied:
Ah, right. Mixed up the logic.

Mic92 reviewed (approved):
Looks good to me otherwise.

JohnRTitor commented on pkgs/build-support/rust/build-rust-package/default.nix:
```diff
@@ -36,6 +37,7 @@
 , cargoDepsHook ? ""
 , buildType ? "release"
 , meta ? {}
+, useCustomVendorFetcher ? false
```
`useAlternateFetcher`?

Mic92 replied:
```suggestion
, usefetchCargoVendor ? false
```

Mic92 commented on pkgs/development/tools/rust/cargo-shuttle/default.nix:
```diff
@@ -20,15 +20,8 @@ rustPlatform.buildRustPackage rec {
-      "permit-pdp-client-rs-0.2.0" = "sha256-F9wSvo3WzoRXjZb+We0Bvcwx3rRSG1QxXPsvrmtIN38=";
-    };
-  };
+  useCustomVendorFetcher = true;
```
```suggestion
  useFetchCargoVendor = true;
```

Mic92 commented:
Ah yeah and to be in line with rust's best practices, could you also drop a paragraph in `doc/languages-frameworks/rust.section.md`? Sorry for the incremental review.

JohnRTitor commented:
> Ah yeah and to be in line with rust's best practices, could you also drop a paragraph in `doc/languages-frameworks/rust.section.md`?

Perhaps a release notes as well?

TomaSajt commented:
@emilazy IIRC you wanted to work on a very orthogonal alternative to this.
What do you think, should we start using this before your work is completed?

JohnRTitor commented on pkgs/development/compilers/rust/make-rust-platform.nix:
```diff
@@ -18,9 +18,11 @@ rec {
     inherit cargo;
   };
 
+  fetchCargoVendor = buildPackages.callPackage ../../../build-support/rust/fetch-cargo-vendor.nix { inherit cargo; };
```
fetchCargoVendor sounds about right because it's is fetching the crate deps as well as git deps.

`fetchCargoGitDeps` could work but again, we are fetching the crate deps as well.

Mic92 replied:
I think `fetchCargoVendor` good enough.

JohnRTitor reviewed (commented):
Could you add docs for it in the rust docs page?

TomaSajt committed:
docs: mention fetchCargoVendor

TomaSajt commented:
Not the best at writing docs, but I added some sections mentioning the changes.
Also, I modified the error message that shows up when trying to use `fetchCargoTarball`.

emilazy commented:
> @emilazy IIRC you wanted to work on a very orthogonal alternative to this.
> What do you think, should we start using this before your work is completed?

Hi! Yes, I still plan to work on https://github.com/NixOS/nixpkgs/issues/333702 for 25.05. However I’ve been watching this work with interest too and think it’s a great improvement – having to vendor for Git hashes is painful, and even if Nixpkgs uses a scheme like my proposal, this would still have use for out‐of‐tree users. So I think it’d be great to land this and prefer it over vendoring more `Cargo.lock` files just because of Git dependencies!

I think it might not sense to do a big treewide migration of existing Rust packages to use it if we’re going to turn around and migrate them to something with stronger properties (supporting easy patching of individual crates in one place, more reliable devendoring, insight into dependency graphs for supply chain security, etc.), but obviously my proposal still needs me to actually prototype it out, confirm that it works well, and refine it to the point where it can be merged into the tree before we could even start on that, so I don’t want to tell people how to spend their time or block incremental improvements at all :)

TomaSajt commented:
Gathered some data:

What makes this a bit inaccurate is that I assume that, if it uses `outputHashes`, then the lockfile can be removed from tree. This is not always the case: it still might be that upstream doesn't have a lockfile committed.

```
$ find -name "Cargo.lock" | wc -l
365
$ find -name "Cargo.lock" | xargs cat | wc -l
1248710

$ find $(rg outputHashes -l | xargs dirname) -name "Cargo.lock" | wc -l
294
$ find $(rg outputHashes -l | xargs dirname) -name "Cargo.lock" | xargs cat | wc -l
1139579
```
so around 90% of the total lines is from lockfiles that contain git deps.

(also, ~1million negative line diff would pretty crazy to do)

```
$ du pkgs/ -hd0
392M

$ find $(rg outputHashes -l | xargs dirname) -name "Cargo.lock" -delete

$ du pkgs/ -hd0
365M
```
meaning that these removable lockfiles took up around 7% of the space in the `pkgs` directory.

Looking at the tarball sizes before and after deletion:
`44.9M -> 38.3M`

JohnRTitor reviewed (approved):
LGTM, will merge if no objections

TomaSajt commented:
Both fetchers are now passed the same args other than `sha256` and `cargoUpdateHook`
(I'm not sure where `cargoUpdateHook` would need to be used, so I left it out)

JohnRTitor commented:
As far as my understanding of the codebase, `cargoUpdateHook` is a way to inject a [bash snippet to the buildPhase](https://github.com/NixOS/nixpkgs/blob/7e3ceaa530f63ceff1e894a08050dc89a7eaff9d/pkgs/build-support/rust/fetch-cargo-tarball/default.nix#L114).

Since the entire script is written in Python, and [in the build phase](https://github.com/NixOS/nixpkgs/pull/349360/files#diff-b0fea804753d211e8b2f61a5f88f98a6b094a1db5752a86dbcfc9910fe4665c8R65) the python script is just called, I don't think we have a much of a choice here. Injecting Python isn't as easy as bash.

But I'd let one of the @NixOS/rust maintainers decide this.

TomaSajt commented:
Seems like https://github.com/NixOS/nixpkgs/pull/35377 added it back this functionality after it was removed after a big refactor.
I don't think we should care about this too much.

JohnRTitor commented:
Agreed, let's hold back this no longer and merge this after eval passes.

Adding `cargoUpdateHook` is not feasible at this stage.

JohnRTitor commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 349360`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 14 packages built:</summary>
  <ul>
    <li>cargo-shuttle</li>
    <li>hieroglyphic</li>
    <li>mitmproxy2swagger</li>
    <li>mitmproxy2swagger.dist</li>
    <li>nixpkgs-manual</li>
    <li>popsicle</li>
    <li>python311Packages.json-stream</li>
    <li>python311Packages.json-stream-rs-tokenizer</li>
    <li>python311Packages.json-stream-rs-tokenizer.dist</li>
    <li>python311Packages.json-stream.dist</li>
    <li>python312Packages.json-stream</li>
    <li>python312Packages.json-stream-rs-tokenizer</li>
    <li>python312Packages.json-stream-rs-tokenizer.dist</li>
    <li>python312Packages.json-stream.dist</li>
  </ul>
</details>

Atemu commented:
> Nothing is stopping us from also compressing the final vendor directory into a tarball, just like with `fetchCargoTarball`.
> Tarballs - obviously - take up less space, so ideally we would do this.

I don't think we should do that btw. Handling plain files is simpler and faster.

If you cared about space usage of intermediate build files etc. you should just use a filesystem that supports transparent compression. It's a >50% reduction in space usage on my Nix store with only zstd:1.  
You'd also simply GC if you need space anyways which would clean up this build-time dep in any case.

> The rebuilds of the vendor directories would not take much time, but the used up space would be a concern.

It's actually slightly worse for hydra/cache.nixos.org as it has to cache this for every system separately, so take everything here x4. And of course rebuilds, so multiply by ~20-30 per year.

For "home users" the same argument w.r.t. space use applies here though. Additionally, this and the hydra issue could be remedied by unpacking the tarballs in the stable FOD format and them simply symlinking the actual sources into the unstable format. That way the "heavy-weight" ground truth data is in the FOD while the system- and python-specific derivations are just a light-weight translation layers. 
This might be something to consider for a V2.

This would also protect us against crates.io possibly changing their gzip implementation or version of downloaded tarballs. I don't know whether there is any "contract" around that. This has bitten us with GitHub tarballs before for instance.

> It would be great if we could tell hydra not to cache certain derivations.

That's not possible with the current implementation of nix remote builds AFAIK.

TomaSajt commented:
> Tis and the hydra issue could be remedied by unpacking the tarballs in the stable FOD format and them simply symlinking the actual sources into the unstable format. That way the "heavy-weight" ground truth data is in the FOD while the system- and python-specific derivations are just a light-weight translation layers. 
> This might be something to consider for a V2.
> 
> This would also protect us against crates.io possibly changing their gzip implementation or version of downloaded tarballs. I don't know whether there is any "contract" around that. This has bitten us with GitHub tarballs before for instance.

The cargo tarballs downloaded are checksummed, so it should be fine.

I did not think of the symlinking option, seems like a promising idea. The unmodified stuff would be symlinks (non-git) and the normalized stuff would be just a copy (git with workspace stuff).

Though, yes, that woukd require unpacking in the fod, which I was afraid to do, but should be viable.

Atemu commented:
> that would require unpacking in the fod, which I was afraid to do

Is there any particular reason you were "afraid" to do it?

TomaSajt commented:
> > that would require unpacking in the fod, which I was afraid to do
> 
> Is there any particular reason you were "afraid" to do it?

I wanted to do the minimun possible processing in the staging phase, to lower the possibility of messing something up in the implementation.

TomaSajt commented:
Also, importCargoLock should also "suffer" from the same storage problem, since it doesn't use its internal FODs directly either.
(for git, it's workspace normalization, using python) (for non-git, it just uses tar, so it should encounter less rebuilds).

This means that fetchCargoVendor should not be worsening the storage situation by much, compared to importCargoLock

Atemu commented:
> I wanted to do the minimum possible processing in the staging phase, to lower the possibility of messing something up in the implementation.

I see. That's reasonable but I also think unpacking a tarball is hard to get wrong. ([If you remember the command](https://xkcd.com/1168/))

> importCargoLock should also "suffer" from the same storage problem, since it doesn't use its internal FODs directly either

Indeed. Given that it has [more pressing issues](https://github.com/NixOS/nixpkgs/issues/327063), I don't think we should place much importance on it for sustainable use within Nixpkgs.

> This means that fetchCargoVendor should not be worsening the storage situation by much, compared to importCargoLock

You are right of course. This is merely a low-hanging fruit optimisation beyond the status quo.

I'll open a separate issue to discuss optimisations for `fetchCargoVendor`.

getchoo commented:
Making a backport as this is going to prevent a lot of other backports from going through, and shouldn't be a breaking change in any way. It can also lead to some nasty surprises on stable, as only setting `useFetchCargoVendor = true` won't cause an evaluation failure about `fetchCargoVendor` missing (found this in https://github.com/NixOS/nixpkgs/pull/362266)

jian-lin commented:
I created https://github.com/NixOS/nixpkgs/issues/377558 to track the support of other registries.
